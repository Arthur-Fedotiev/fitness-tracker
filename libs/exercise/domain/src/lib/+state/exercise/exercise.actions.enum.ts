export enum EXERCISES_ACTION_NAMES {
  'LOAD_EXERCISES' = '[Exercises/API] LOAD_EXERCISES',
  'LOAD_EXERCISES_SUCCESS' = '[Exercises/API] LOAD_EXERCISES_SUCCESS',
  'LOAD_EXERCISES_FAILURE' = '[Exercises/API] LOAD_EXERCISES_FAILURE',

  'FIND_EXERCISES' = '[Exercises/API] FIND_EXERCISES',
  'FIND_EXERCISES_SUCCESS' = '[Exercises/API] FIND_EXERCISES_SUCCESS',
  'FIND_EXERCISES_FAILURE' = '[Exercises/API] FIND_EXERCISES_FAILURE',

  'REFRESH_EXERCISES' = '[Exercises/API] REFRESH_EXERCISES',
  'REFRESH_EXERCISES_SUCCESS' = '[Exercises/API] REFRESH_EXERCISES_SUCCESS',
  'REFRESH_EXERCISES_FAILURE' = '[Exercises/API] REFRESH_EXERCISES_FAILURE',

  'EMPTY_EXERCISES' = '[Exercises/Release Resource] EMPTY_EXERCISES',

  'EXERCISE_SAVED' = '[Exercises/ExerciseDetails] EXERCISE_SAVED',
  'EXERCISE_SAVED_SUCCESS' = '[Exercises/ExerciseDetails] EXERCISE_SAVED_SUCCESS',
  'EXERCISE_SAVED_FAILURE' = '[Exercises/ExerciseDetails] EXERCISE_SAVED_FAILURE',

  'DELETE_EXERCISE' = '[Exercises/API] DELETE_EXERCISE',
  'DELETE_EXERCISE_SUCCESS' = '[Exercises/API] DELETE_EXERCISE_SUCCESS',
  'DELETE_EXERCISE_FAILURE' = '[Exercises/API] DELETE_EXERCISE_FAILURE',

  'LOAD_EXERCISE_DETAILS' = '[Exercises/API] LOAD_EXERCISE_DETAILS',
  'LOAD_EXERCISE_DETAILS_SUCCESS' = '[Exercises/API] LOAD_EXERCISE_DETAILS_SUCCESS',
  'LOAD_EXERCISE_DETAILS_FAILURE' = '[Exercises/API] LOAD_EXERCISE_DETAILS_FAILURE',
  'RELEASE_EXERCISE_DETAILS' = '[Exercises/Release Resource] RELEASE_EXERCISE_DETAILS',

  'OPEN_EXERCISE_DETAILS_DIALOG' = '[Exercises/UI] OPEN_EXERCISE_DETAILS_DIALOG',
}
