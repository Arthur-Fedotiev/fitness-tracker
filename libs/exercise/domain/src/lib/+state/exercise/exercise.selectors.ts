import { createFeatureSelector, createSelector } from '@ngrx/store';
import {
  EXERCISE_FEATURE_KEY,
  State,
  ExercisePartialState,
  exerciseAdapter,
} from './exercise.reducer';

// Lookup the 'Exercise' feature state managed by NgRx
export const getExerciseState = createFeatureSelector<
  ExercisePartialState,
  State
>(EXERCISE_FEATURE_KEY);

const { selectAll, selectEntities } = exerciseAdapter.getSelectors();

export const getExerciseLoaded = createSelector(
  getExerciseState,
  (state: State) => state.loaded,
);

export const getExerciseError = createSelector(
  getExerciseState,
  (state: State) => state.error,
);

export const getAllExercise = createSelector(getExerciseState, (state: State) =>
  selectAll(state),
);

export const getExerciseEntities = createSelector(
  getExerciseState,
  (state: State) => selectEntities(state),
);

export const getSelectedId = createSelector(
  getExerciseState,
  (state: State) => state.selectedId,
);

export const getSelected = createSelector(
  getExerciseEntities,
  getSelectedId,
  (entities, selectedId) => selectedId && entities[selectedId],
);
