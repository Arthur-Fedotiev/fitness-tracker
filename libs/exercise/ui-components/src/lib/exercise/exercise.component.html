<mat-card class="card mat-elevation-z8" tabindex="0">
  <mat-card-title-group>
    <img
      class="card__avatar"
      mat-card-sm-image
      [src]="exercise.avatarUrl"
      ftWithFallback
    />
    <img
      fxHide.lt-sm
      class="card__avatar"
      mat-card-sm-image
      [src]="exercise.avatarSecondaryUrl"
      ftWithFallback
    />
    <mat-card-title class="card__title">
      <span
        ngClass.sm="ft-sm"
        ngClass.md="ft-md"
        ngClass.lg="ft-lg"
        ngClass.xl="ft-xl"
        >{{ exercise.name }}</span
      >
    </mat-card-title>
  </mat-card-title-group>
  <mat-card-content>
    <div class="subtitle-item">
      <p translate>EXERCISE.targetMuscle</p>
      <p>{{ 'MUSCLES.' + exercise.targetMuscle | translate }}</p>
    </div>
    <div class="subtitle-item">
      <p translate>EXERCISE.equipment</p>
      <p>{{ 'EQUIPMENT.' + exercise.equipment | translate }}</p>
    </div>
    <div class="subtitle-item">
      <p translate>EXERCISE.exerciseType</p>
      <p>{{ 'EXERCISE-TYPES.' + exercise.exerciseType | translate }}</p>
    </div>
  </mat-card-content>
  <mat-card-actions class="exercise-actions">
    <button
      mat-button
      class="mat-raised-button mat-primary"
      (click)="viewExercise(exercise)"
    >
      {{ 'EXERCISE.viewBtn' | translate }}
    </button>

    <button
      *ftRoles="roles.ADMIN"
      mat-button
      class="mat-raised-button mat-accent"
      (click)="editExercise(exercise)"
    >
      {{ 'EXERCISE.editBtn' | translate }}
    </button>
    <button
      *ftRoles="roles.ADMIN"
      mat-mini-fab
      color="warn"
      (click)="deleteExercise(exercise)"
    >
      <mat-icon>delete</mat-icon>
    </button>
    <button
      *ftRoles="roles.ADMIN"
      mat-mini-fab
      class="add-btn"
      color="success"
      (click)="addedToWorkout(exercise)"
    >
      <mat-icon>add</mat-icon>
    </button>
  </mat-card-actions>
</mat-card>
