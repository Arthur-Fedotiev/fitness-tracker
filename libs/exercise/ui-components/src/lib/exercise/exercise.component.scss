.exercise-actions {
  display: flex;
  gap: 1rem;
  margin: 0 auto;
  text-align: center;
}

.add-btn {
  margin-left: 0.5rem;
}

.card {
  border-radius: 1.5rem;
  &__avatar {
    border-radius: 0.5rem;
  }
  &__title {
    font-size: 1rem;
    text-align: center;
  }

  .subtitle-item > p:first-child {
    color: #ff5722;
  }
}

:host ::ng-deep {
  mat-card-title-group {
    display: flex;
    align-items: center;
    & > div {
      flex: 0.9;
    }
  }

  mat-card-content {
    display: flex;
    flex-flow: row nowrap;
    justify-content: space-between;

    .subtitle-item {
      margin-top: 0.5rem;
      :first-child {
        font-size: 1rem;
        font-weight: 500;
        white-space: nowrap;
      }
    }
  }
}
