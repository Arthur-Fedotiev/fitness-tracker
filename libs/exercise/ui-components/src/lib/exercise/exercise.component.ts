import {
  Component,
  ChangeDetectionStrategy,
  EventEmitter,
  Input,
  Output,
} from '@angular/core';
import { ROLES } from 'shared-package';
import { ExerciseVM } from '../models/exercise-vm';
import { MatIconModule } from '@angular/material/icon';
import { MatButtonModule } from '@angular/material/button';
import { TranslateModule } from '@ngx-translate/core';
import { ExtendedModule } from '@angular/flex-layout/extended';
import { NgOptimizedImage, TitleCasePipe, NgIf } from '@angular/common';
import { MatCardModule } from '@angular/material/card';
import {
  ImgFallbackDirective,
  E2eDirective,
  EXERCISE_AVATAR_FALLBACK_IMG,
  EXERCISE_AVATAR_FALLBACK_SECONDARY_IMG,
} from '@fitness-tracker/shared/utils';

@Component({
  selector: 'components-exercise',
  templateUrl: './exercise.component.html',
  styleUrls: ['./exercise.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  standalone: true,
  imports: [
    MatCardModule,
    ImgFallbackDirective,
    NgOptimizedImage,
    NgIf,
    ExtendedModule,
    E2eDirective,
    TranslateModule,
    MatButtonModule,
    MatIconModule,
    TitleCasePipe,
  ],
})
export class ExerciseComponent {
  @Input()
  public exercise!: ExerciseVM;
  @Input()
  public idx!: number;

  @Output()
  public readonly exerciseEdited = new EventEmitter<string>();
  @Output()
  public readonly exerciseViewed = new EventEmitter<string>();
  @Output()
  public readonly exerciseDeleted = new EventEmitter<string>();
  @Output()
  public readonly exerciseAddedToWorkout = new EventEmitter<string>();

  protected readonly roles = ROLES;
  protected readonly avatarFallback = EXERCISE_AVATAR_FALLBACK_IMG;
  protected readonly secondaryAvatarFallback =
    EXERCISE_AVATAR_FALLBACK_SECONDARY_IMG;

  public viewExercise({ id }: ExerciseVM): void {
    this.exerciseViewed.emit(id);
  }

  public editExercise({ id }: ExerciseVM): void {
    this.exerciseEdited.emit(id);
  }

  public deleteExercise({ id }: ExerciseVM): void {
    this.exerciseDeleted.emit(id);
  }

  public addedToWorkout({ id }: ExerciseVM): void {
    this.exerciseAddedToWorkout.emit(id);
  }
}
