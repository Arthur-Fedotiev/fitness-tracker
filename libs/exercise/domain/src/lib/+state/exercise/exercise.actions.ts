import { createAction, props } from '@ngrx/store';
import { Exercise } from '../../entities/exercise';

export const loadExercise = createAction('[Exercise] Load Exercise');

export const loadExerciseSuccess = createAction(
  '[Exercise] Load Exercise Success',
  props<{ exercise: Exercise[] }>(),
);

export const loadExerciseFailure = createAction(
  '[Exercise] Load Exercise Failure',
  props<{ error: any }>(),
);
