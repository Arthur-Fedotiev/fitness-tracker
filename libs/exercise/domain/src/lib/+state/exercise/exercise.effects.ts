import { Inject, Injectable, Type } from '@angular/core';
import { createEffect, Actions, ofType } from '@ngrx/effects';

import * as ExercisesActions from './exercise.actions';
import { EXERCISES_ACTION_NAMES } from './exercise.actions.enum';
import {
  catchError,
  concatMap,
  forkJoin,
  map,
  mergeMap,
  Observable,
  of,
  switchMap,
  tap,
  withLatestFrom,
} from 'rxjs';

import { WithPayload } from '@fitness-tracker/shared/utils';
import { Action, Store } from '@ngrx/store';
import { LanguageCodes } from 'shared-package';
import { TypedAction } from '@ngrx/store/src/models';
import { MatDialog } from '@angular/material/dialog';
import { first, switchMapTo } from 'rxjs/operators';
import { FirebaseExerciseDataService } from '../../infrastructure/exercise.data.service';
import { selectLanguage } from '@fitness-tracker/shared/data-access';
import {
  GetExerciseRequestDto,
  SearchOptions,
} from '../../entities/dto/request/get/get-exercise-request.dto';
import { ExerciseResponseDto } from '../../entities/dto/response/exercise-response.dto';
import { CreateUpdateExerciseRequestDTO } from '../../entities/dto/request/update/exercise-create-update-request.dto';
import { ExerciseDetailsDialogComponent } from '../../application/providers/exercise-details-dialog.provider';

@Injectable()
export class ExerciseEffects {
  public readonly findExercises$ = createEffect(() =>
    this.actions$.pipe(
      ofType(
        EXERCISES_ACTION_NAMES.FIND_EXERCISES,
        EXERCISES_ACTION_NAMES.REFRESH_EXERCISES,
      ),
      withLatestFrom(this.store.select(selectLanguage)),
      concatMap(
        ([{ payload, type }, language]: [
          TypedAction<
            | EXERCISES_ACTION_NAMES.FIND_EXERCISES
            | EXERCISES_ACTION_NAMES.REFRESH_EXERCISES
          > &
            WithPayload<SearchOptions>,
          LanguageCodes,
        ]) =>
          this.exercisesService
            .findExercises(
              new GetExerciseRequestDto(
                { ...payload },
                language,
                type === EXERCISES_ACTION_NAMES.REFRESH_EXERCISES,
              ),
            )
            .pipe(
              tap((exercises) => {
                console.log(exercises);
              }),
              map((exercises: ExerciseResponseDto[]) =>
                type === EXERCISES_ACTION_NAMES.REFRESH_EXERCISES
                  ? ExercisesActions.refreshExercisesSuccess({
                      payload: exercises,
                    })
                  : ExercisesActions.findExercisesSuccess({
                      payload: { exercises, firstPage: !!payload.firstPage },
                    }),
              ),
              catchError(() => of(ExercisesActions.findExercisesFailure())),
            ),
      ),
    ),
  );

  public readonly saveExercise$ = createEffect(() =>
    this.actions$.pipe(
      ofType(EXERCISES_ACTION_NAMES.EXERCISE_SAVED),
      mergeMap(({ payload }: WithPayload<CreateUpdateExerciseRequestDTO>) =>
        this.exercisesService.createOrUpdateExercise(payload).pipe(
          map(() => ExercisesActions.exerciseSavedSuccess()),
          catchError(() => [ExercisesActions.exerciseSavedFailure()]),
        ),
      ),
    ),
  );

  public readonly deleteExercise$ = createEffect(() =>
    this.actions$.pipe(
      ofType(EXERCISES_ACTION_NAMES.DELETE_EXERCISE),
      mergeMap(
        ({ payload }: WithPayload<string>): Observable<Action> =>
          this.exercisesService.deleteExercise(payload).pipe(
            map(() => ExercisesActions.deleteExerciseSuccess({ payload })),
            catchError(() => of(ExercisesActions.deleteExerciseFailure())),
          ),
      ),
    ),
  );

  public readonly loadExerciseDetails$ = createEffect(() =>
    this.actions$.pipe(
      ofType(EXERCISES_ACTION_NAMES.LOAD_EXERCISE_DETAILS),
      withLatestFrom(this.store.select(selectLanguage)),
      switchMap(
        ([{ payload }, lang]: [
          WithPayload<string>,
          LanguageCodes,
        ]): Observable<Action> =>
          this.exercisesService.getExerciseDetails(payload, lang).pipe(
            map((payload: ExerciseResponseDto) =>
              ExercisesActions.loadExerciseDetailsSuccess({ payload }),
            ),
            catchError(() => of(ExercisesActions.loadExerciseDetailsFailure())),
          ),
      ),
    ),
  );

  public readonly showExerciseDetails$ = createEffect(() =>
    this.actions$.pipe(
      ofType(EXERCISES_ACTION_NAMES.OPEN_EXERCISE_DETAILS_DIALOG),
      map(({ payload }) => ExercisesActions.loadExerciseDetails({ payload })),
    ),
  );

  public readonly openExerciseDetailsDialog$ = createEffect(
    () =>
      this.actions$.pipe(
        ofType(EXERCISES_ACTION_NAMES.OPEN_EXERCISE_DETAILS_DIALOG),
        switchMapTo(
          this.actions$.pipe(
            ofType(
              EXERCISES_ACTION_NAMES.LOAD_EXERCISE_DETAILS_SUCCESS,
              EXERCISES_ACTION_NAMES.LOAD_EXERCISE_DETAILS_FAILURE,
            ),
            first(),
          ),
        ),
        switchMap(({ payload }: WithPayload<ExerciseResponseDto>) =>
          forkJoin([of(payload), this.detailsDialogFactory]),
        ),
        tap(([exercise, component]) =>
          this.dialog.open(component, { data: { exercise } }),
        ),
      ),
    { dispatch: false },
  );

  constructor(
    @Inject(ExerciseDetailsDialogComponent)
    private readonly detailsDialogFactory: Observable<Type<unknown>>,
    private readonly actions$: Actions,
    private readonly exercisesService: FirebaseExerciseDataService,
    private readonly store: Store,
    private readonly dialog: MatDialog,
  ) {}
}
