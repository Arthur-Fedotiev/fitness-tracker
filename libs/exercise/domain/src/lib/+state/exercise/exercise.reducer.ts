import { EntityState, EntityAdapter, createEntityAdapter } from '@ngrx/entity';
import { createReducer, on, Action } from '@ngrx/store';
import * as ExerciseActions from './exercise.actions';

import { ExerciseResponseDto } from '../../entities/dto/response/exercise-response.dto';

export const EXERCISES_FEATURE_KEY = 'exercises';
export const sortComparer = (a: ExerciseResponseDto, b: ExerciseResponseDto) =>
  b.rating - a.rating > 0 ? 1 : -1;

export interface State extends EntityState<ExerciseResponseDto> {
  selectedId?: string | number;
  selectedExercise: ExerciseResponseDto | null;
  loading: boolean;
  error?: string | null;
}

export interface ExercisesPartialState {
  readonly [EXERCISES_FEATURE_KEY]: State;
}

export const exercisesAdapter: EntityAdapter<ExerciseResponseDto> =
  createEntityAdapter<ExerciseResponseDto>({
    sortComparer,
  });

export const initialState: State = exercisesAdapter.getInitialState({
  loading: false,
  selectedExercise: null,
  collectionsMeta: null,
});

const exercisesReducer = createReducer(
  initialState,
  on(ExerciseActions.refreshExercises, (state) => ({
    ...state,
    loading: true,
    error: null,
  })),
  on(ExerciseActions.refreshExercisesSuccess, (state, { payload }) =>
    exercisesAdapter.setAll(payload, { ...state, loading: false }),
  ),
  on(ExerciseActions.refreshExercisesFailure, (state) => ({
    ...state,
    loading: false,
  })),
  on(ExerciseActions.findExercises, (state) => ({
    ...state,
    loading: true,
    error: null,
  })),
  on(ExerciseActions.findExercisesSuccess, (state, { payload }) => {
    const adapterAction = payload.firstPage
      ? exercisesAdapter.setAll
      : exercisesAdapter.addMany;

    return adapterAction(payload.exercises, {
      ...state,
      loading: false,
    });
  }),
  on(ExerciseActions.findExercisesFailure, (state) => ({
    ...state,
    loading: false,
  })),
  on(ExerciseActions.emptyExercisesList, (state) =>
    exercisesAdapter.removeAll(state),
  ),
  on(ExerciseActions.createExercise, (state, { payload }) =>
    exercisesAdapter.addOne(payload, { ...state, loading: false }),
  ),
  on(ExerciseActions.deleteExerciseSuccess, (state, { payload }) =>
    exercisesAdapter.removeOne(payload, state),
  ),
  on(ExerciseActions.loadExerciseDetails, (state) => ({
    ...state,
    loading: true,
  })),
  on(ExerciseActions.loadExerciseDetailsSuccess, (state, { payload }) => ({
    ...state,
    loading: false,
    selectedExercise: payload,
  })),
  on(ExerciseActions.loadExerciseDetailsFailure, (state) => ({
    ...state,
    loading: false,
  })),
  on(ExerciseActions.releaseExerciseDetails, (state) => ({
    ...state,
    selectedExercise: null,
  })),
);

export function reducer(state: State | undefined, action: Action) {
  return exercisesReducer(state, action);
}
