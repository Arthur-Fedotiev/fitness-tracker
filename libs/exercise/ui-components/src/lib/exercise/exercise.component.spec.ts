import { CommonModule } from '@angular/common';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { FlexLayoutModule } from '@angular/flex-layout';
import { AUTH_FACADE_MOCK_PROVIDER } from '@fitness-tracker/auth/data';
import { RolesModule } from '@fitness-tracker/auth/feature';
import { MaterialModule } from '@fitness-tracker/shared-ui-material';
import { TranslateTestingModule } from '@fitness-tracker/shared/testing';
import { ImgFallbackModule } from '@fitness-tracker/shared/utils';

import { ExerciseComponent } from './exercise.component';

describe('ExerciseComponent', () => {
  let component: ExerciseComponent;
  let fixture: ComponentFixture<ExerciseComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        CommonModule,
        MaterialModule,
        FlexLayoutModule,
        ImgFallbackModule,
        RolesModule,
        TranslateTestingModule,
      ],
      providers: [AUTH_FACADE_MOCK_PROVIDER],
      declarations: [ExerciseComponent],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(ExerciseComponent);
    component = fixture.componentInstance;

    component.exercise = {
      id: '1',
      name: 'name',
      avatarSecondaryUrl: 'avatarSecondaryUrl',
      avatarUrl: 'avatarUrl',
      coverUrl: 'coverUrl',
      equipment: 'equipment',
      targetMuscle: 'targetMuscle',
      exerciseType: 'exerciseType',
    };

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should render correct markup', () => {
    expect(fixture.nativeElement).toMatchSnapshot();
  });

  describe('emit events', () => {
    const verifyEmitEvent = (eventHandler: string, emitterName: string) => {
      const spy = jest.spyOn(component[emitterName], 'emit');

      component[eventHandler]({ id: '1' });

      expect(spy).toHaveBeenCalledWith('1');
    };

    const eventsTestingData = [
      ['editExercise', 'exerciseEdited'],
      ['deleteExercise', 'exerciseDeleted'],
      ['viewExercise', 'exerciseViewed'],
      ['addedToWorkout', 'exerciseAddedToWorkout'],
    ];

    it.each(eventsTestingData)(
      'should emit %s event',
      (eventHandler, emitterName) => {
        verifyEmitEvent(eventHandler, emitterName);
      },
    );
  });
});
