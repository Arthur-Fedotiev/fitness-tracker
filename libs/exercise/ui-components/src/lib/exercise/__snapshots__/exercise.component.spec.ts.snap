// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExerciseComponent should render correct markup 1`] = `
<div
  id="root1"
>
  <mat-card
    class="mat-card mat-focus-indicator card mat-elevation-z8"
    tabindex="0"
  >
    <mat-card-title-group
      class="mat-card-title-group"
    >
      <div>
        <mat-card-title
          class="mat-card-title card__title"
        >
          <span
            ngclass.lg="ft-lg"
            ngclass.md="ft-md"
            ngclass.sm="ft-sm"
            ngclass.xl="ft-xl"
          >
            name
          </span>
        </mat-card-title>
      </div>
      <img
        class="mat-card-sm-image card__avatar"
        ftwithfallback=""
        mat-card-sm-image=""
        src="avatarUrl"
      />
      <img
        class="mat-card-sm-image card__avatar"
        ftwithfallback=""
        fxhide.lt-sm=""
        mat-card-sm-image=""
        src="avatarSecondaryUrl"
      />
    </mat-card-title-group>
    <mat-card-content
      class="mat-card-content"
    >
      <div
        class="subtitle-item"
      >
        <p
          translate=""
        >
          EXERCISE.targetMuscle
        </p>
        <p>
          MUSCLES.targetMuscle
        </p>
      </div>
      <div
        class="subtitle-item"
      >
        <p
          translate=""
        >
          EXERCISE.equipment
        </p>
        <p>
          EQUIPMENT.equipment
        </p>
      </div>
      <div
        class="subtitle-item"
      >
        <p
          translate=""
        >
          EXERCISE.exerciseType
        </p>
        <p>
          EXERCISE-TYPES.exerciseType
        </p>
      </div>
    </mat-card-content>
    <mat-card-actions
      class="mat-card-actions exercise-actions"
    >
      <button
        class="mat-focus-indicator mat-raised-button mat-primary mat-button mat-button-base"
        mat-button=""
      >
        <span
          class="mat-button-wrapper"
        >
           EXERCISE.viewBtn 
        </span>
        <span
          class="mat-ripple mat-button-ripple"
          matripple=""
        />
        <span
          class="mat-button-focus-overlay"
        />
      </button>
      <button
        class="mat-focus-indicator mat-raised-button mat-accent mat-button mat-button-base"
        mat-button=""
      >
        <span
          class="mat-button-wrapper"
        >
           EXERCISE.editBtn 
        </span>
        <span
          class="mat-ripple mat-button-ripple"
          matripple=""
        />
        <span
          class="mat-button-focus-overlay"
        />
      </button>
      
      <button
        class="mat-focus-indicator mat-mini-fab mat-button-base mat-warn"
        color="warn"
        mat-mini-fab=""
      >
        <span
          class="mat-button-wrapper"
        >
          <mat-icon
            aria-hidden="true"
            class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
            data-mat-icon-type="font"
            role="img"
          >
            delete
          </mat-icon>
        </span>
        <span
          class="mat-ripple mat-button-ripple mat-button-ripple-round"
          matripple=""
        />
        <span
          class="mat-button-focus-overlay"
        />
      </button>
      
      <button
        class="mat-focus-indicator add-btn mat-mini-fab mat-button-base mat-success"
        color="success"
        mat-mini-fab=""
      >
        <span
          class="mat-button-wrapper"
        >
          <mat-icon
            aria-hidden="true"
            class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
            data-mat-icon-type="font"
            role="img"
          >
            add
          </mat-icon>
        </span>
        <span
          class="mat-ripple mat-button-ripple mat-button-ripple-round"
          matripple=""
        />
        <span
          class="mat-button-focus-overlay"
        />
      </button>
      
    </mat-card-actions>
  </mat-card>
</div>
`;
